numba/cpython/randomimpl.py

```diff
@@ -1304,66 +1539,91 @@ def _impl(n, p):
-        sig = signature(sig.return_type, types.float64)
-        args = (ir.Constant(double, 1.0),)
-
-    ret = context.compile_internal(builder, poisson_impl, sig, args)
-    builder.store(ret, retptr)
-    builder.branch(bbend)
-    builder.position_at_end(bbend)
-    res = builder.load(retptr)
-    return impl_ret_untracked(context, builder, sig.return_type, res)
+@overload(np.random.poisson)
+def poisson_impl0():
+    return lambda: np.random.poisson(1.0)
+
+
+@overload(np.random.poisson)
+def poisson_impl1(lam):
+    if isinstance(lam, types.Number):
+        @intrinsic
+        def _impl(typingcontext, lam):
+            lam_preprocessor = _double_preprocessor(lam)
+
+            def codegen(context, builder, sig, args):
+                state_ptr = get_np_state_ptr(context, builder)
+
+                retptr = cgutils.alloca_once(builder, int64_t, name="ret")
+                bbcont = builder.append_basic_block("bbcont")
+                bbend = builder.append_basic_block("bbend")
+
+                lam, = args
+                lam = lam_preprocessor(builder, lam)
+                big_lam = builder.fcmp_ordered('>=', lam, ir.Constant(double, 10.0))
+                with builder.if_then(big_lam):
+                    # For lambda >= 10.0, we switch to a more accurate
+                    # algorithm (see _random.c).
+                    fnty = ir.FunctionType(int64_t, (rnd_state_ptr_t, double))
+                    fn = cgutils.get_or_insert_function(builder.function.module,
+                                                        fnty,
+                                                        "numba_poisson_ptrs")
+                    ret = builder.call(fn, (state_ptr, lam))
+                    builder.store(ret, retptr)
+                    builder.branch(bbend)
+
+                builder.branch(bbcont)
+                builder.position_at_end(bbcont)
+
+                _random = np.random.random
+                _exp = math.exp
+
+                def poisson_impl(lam):
+                    """Numpy's algorithm for poisson() on small *lam*.
+
+                    This method is invoked only if the parameter lambda of the
+                    distribution is small ( < 10 ). The algorithm used is described
+                    in "Knuth, D. 1969. 'Seminumerical Algorithms. The Art of
+                    Computer Programming' vol 2.
+                    """
+                    if lam < 0.0:
+                        raise ValueError("poisson(): lambda < 0")
+                    if lam == 0.0:
+                        return 0
+                    enlam = _exp(-lam)
+                    X = 0
+                    prod = 1.0
+                    while 1:
+                        U = _random()
+                        prod *= U
+                        if prod <= enlam:
+                            return X
+                        X += 1
+
+                ret = context.compile_internal(builder, poisson_impl, sig, args)
```

Comment:
Not for this PR, but in general, whilst this is a faithful port of the existing code, use of `compile_internal` can and should be avoided. The call should either be resolved via the context or ideally the overload should be entirely rewritten to instead have the overload implementation written as a JIT compilable Python implementation with intrinsic calls where needed.